From db380bdd26566427caa59b5eed4294f11d61746e Mon Sep 17 00:00:00 2001
From: Sebanc <sebanc.xda@gmail.com>
Date: Wed, 9 Oct 2019 21:49:48 +0200
Subject: [PATCH] Core Microsoft Surface devices: Disable power management for
 i2c-hid sensors

i2c-hid sensors on Microsoft Surface core devices appear not to be power manageable.
---
 drivers/hid/i2c-hid/i2c-hid-core.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/hid/i2c-hid/i2c-hid-core.c b/drivers/hid/i2c-hid/i2c-hid-core.c
index 3cde7c1..119f29f 100644
--- a/drivers/hid/i2c-hid/i2c-hid-core.c
+++ b/drivers/hid/i2c-hid/i2c-hid-core.c
@@ -39,6 +39,7 @@
 #include <linux/acpi.h>
 #include <linux/of.h>
 #include <linux/regulator/consumer.h>
+#include <linux/dmi.h>
 
 #include <linux/platform_data/i2c-hid.h>
 
@@ -59,6 +60,25 @@
 #define I2C_HID_PWR_ON		0x00
 #define I2C_HID_PWR_SLEEP	0x01
 
+static const struct dmi_system_id devices[] = 
+	{
+		{
+			.ident = "Microsoft Surface 3",
+			.matches = {
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+				DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface 3"),
+			},
+		},
+		{
+			.ident = "Microsoft Surface Pro 4",
+			.matches = {
+				DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+				DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Pro 4"),
+			},
+		},
+		{}
+	};
+
 /* debug option */
 static bool debug;
 module_param(debug, bool, 0444);
@@ -396,6 +399,9 @@ static int i2c_hid_set_power(struct i2c_client *client, int power_state)
 
 	i2c_hid_dbg(ihid, "%s\n", __func__);
 
+	if (dmi_check_system(devices) && (!strncmp(client->name, "MSHW0030", 8) || !strncmp(client->name, "MSHW0102", 8)))
+		return 0;
+
 	/*
 	 * Some devices require to send a command to wakeup before power on.
 	 * The call will get a return value (EREMOTEIO) but device will be
-- 
2.20.1

